include/circt/Dialect/HW/HWOpInterfaces.td

```diff
@@ -46,6 +46,78 @@ def HWModuleLike : OpInterface<"HWModuleLike"> {
   }];
 }
 
+def HWMutableModuleLike : OpInterface<"HWMutableModuleLike"> {
+  let cppNamespace = "circt::hw";
+  let description = "Provide methods to mutate a module.";
+
+  let methods = [
+    InterfaceMethod<"Return the number of inputs to this module",
+    "unsigned", "getNumInputs", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{
+      return $_op->template getAttrOfType<ArrayAttr>("argNames").size();
+    }]>,
+
+    InterfaceMethod<"Return the number of outputs from this module",
+    "unsigned", "getNumOutputs", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{
+      return $_op->template getAttrOfType<ArrayAttr>("resultNames").size();
+    }]>,
+
+    InterfaceMethod<"Return the names of the inputs this module",
+    "mlir::ArrayAttr", "getArgNames", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{
+      return $_op->template getAttrOfType<ArrayAttr>("argNames");
+    }]>,
+
+    InterfaceMethod<"Return the names of the outputs this module",
+    "mlir::ArrayAttr", "getResultNames", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{
+      return $_op->template getAttrOfType<ArrayAttr>("resultNames");
+    }]>,
```

Comment:
These sound like they shuld go into the `HWModuleLike` interface, since they are not concerned with mutation and useful also for immutable use cases.

Can `HWMutableModuleLike` "inherit" or require `HWModuleLike` to be implemented as well? It would be great if the mutable interfaces were just an extension of the immutable interfaces.

Reply:
I was thinking the same, but I haven’t found a way to do it. That’s why I put these here. If you’re gonna be modifying ports, you’re likely gonna need to know some stuff about them.

Reply:
I asked on Discord and it doesn't look like you indeed cannot.